qiskit/qasm3/__init__.py

```diff
@@ -11,34 +11,71 @@
 # that they have been altered from the originals.
 
 """
-==========================
-Qasm (:mod:`qiskit.qasm3`)
-==========================
+================================
+OpenQASM 3 (:mod:`qiskit.qasm3`)
+================================
 
 .. currentmodule:: qiskit.qasm3
 
-.. autosummary::
-    :toctree: ../stubs/
+Qiskit provides some tools for converting between `OpenQASM 3 <https://openqasm.com>`__
+representations of quantum programs, and the :class:`.QuantumCircuit` class.  These will continue to
+evolve as Qiskit's support for the dynamic-circuit capabilities expressed by OpenQASM 3 increases.
 
-    Exporter
-    dumps
-    dump
+
+Exporting to OpenQASM 3
+=======================
+
+The high-level functions are simply :func:`dump` and :func:`dumps`, which respectively export to a
+file (given as a filename) and to a Python string.
+
+.. autofunction:: dump
+.. autofunction:: dumps
+
+Both of these exporter functions are single-use wrappers around the main :class:`Exporter` class.
+For more complex exporting needs, including dumping multiple circuits in a single session, it may be
+more convenient or faster to use the complete interface.
+
+.. autoclass:: Exporter
+    :members:
+
+All of these interfaces will raise :exc:`QASM3ExporterError` on failure.
+
+.. autoexception:: QASM3ExporterError
+
+
+Importing from OpenQASM 3
+=========================
+
+Currently only two high-level functions are offered, as Qiskit support for importing from OpenQASM 3
+is in its infancy, and the implementation is expected to change significantly.  The two functions
+are :func:`load` and :func:`loads`, which are direct counterparts of :func:`dump` and :func:`dumps`,
+respectively loading a program indirectly from a named file and directly from a given string.
+
+.. note::
+
+    To use either function, the package ``qiskit_qasm3_import`` must be installed.  This can be done
+    by installing Qiskit Terra with the ``qasm3-import`` extra, such as by:
+
+    .. code-block:: text
+
+        pip install qiskit-terra[qasm3-import]
+
+.. autofunction:: load
+.. autofunction:: loads
```

Comment:
I do not have any high-level disagreement with these function names on their own. It is however worth pointing out that this API breaks from the current OpenQASM2 APIs - https://qiskit.org/documentation/stubs/qiskit.qasm.Qasm.html#qiskit.qasm.Qasm. 

I wonder if it might make sense to begin renaming the older API as `qasm2` (with module deprecation paths) and to extend the QuantumCircuit methods for OpenQASM3 support?

This could be follow-up work.

Reply:
While I was on holiday, I actually wrote a whole new OpenQASM 2 parser for Qiskit, with the majority in Rust.  It was mostly for my own learning experience, but since it's like 10x faster than the current Qiskit one (with no external Python dependencies), I might make a PR changing it.

With or without though, though, I think we should move to a new `qasm2` module with the names `load`, `loads`, `dump` and `dumps`.  I'm not keen to add an extra method to `QuantumCircuit` because the class is already _massively_ overloaded - I'd rather we keep things in separate packages.  The `load`/`dump` terminology is the standard Python names (e.g. `pickle.dump`, `json.dump`, etc), and it's consistent with the currently existing `qasm3.dump` and `qpy.dump` as well - Qiskit is already semi-inconsistent, but really it's the old OQ2 handling that's inconsistent with the rest of the package direction.

Reply:
Thanks, Jake, I'd be hesitant to add more work and potential breaks to the existing QASM2 parser for those that are using it given we're planning on migrating away from it but the pathway suggested for the qasm2 module seems good.

Reply:
Yeah, agree that there's no reason to break users' workflows.  My rough intent was to keep `QuantumCircuit.from_qasm_{str,file}`, and just cause them to internally call `qasm2.load` and `qasm2.loads` as appropriate (assuming I do actually publish that parser and it's suitable for Terra).